### PR TITLE
[TIR] Add DeclBuffer IR node and functors

### PR DESCRIPTION
**Pull request by vinx13**
This PR is part of https://github.com/apache/tvm/issues/11627. It added the IR node `DeclBufferNode`, IR functors and TVMScript printer / parser.

cc @tqchen @Lunderberg @Hzfengsy @junrushao1994 @wrongtest-intellif

### CODE REVIEW

**Comment by junrushao1994** [comment]
CC @cyx-6 please review for potential TVMScript syntax change :-)

**Comment by cyx-6** [comment]
> CC @cyx-6 please review for potential TVMScript syntactic/semantic change :-)

LGTM. New parser is compatible with this change, and will be updated soon.

**Review by Hzfengsy — APPROVED** [review]

File: src/target/source/codegen_c.cc [hunk 1]
```diff
@@ -29,6 +29,7 @@
 
 #include "../../arith/pattern_match.h"
 #include "codegen_params.h"
+#include "tvm/tir/stmt.h"
```
Comment (junrushao1994) [inline src/target/source/codegen_c.cc:32]: nit

```suggestion
#include <tvm/tir/stmt.h>
```

**Review by junrushao1994 — APPROVED** [review]
Overall LGTM. A quick question: after this change, the parser will treat `T.decl_buffer` as an IR construct instead of an intrinsic that declares a buffer - is my understanding correct?

**Comment by vinx13** [comment]
yes. `T.buffer_decl` will be deprecated in the future

**Pull request merged**